### PR TITLE
fix:claim cached in pvcontroller is not the newest may cause unexpected issue

### PR DESCRIPTION
**Pull request by xiaopinggoer**
#### Which issue(s) this PR fixes:
What this PR does / why we need it:
fix issue: https://github.com/kubernetes/kubernetes/issues/105131

Which issue(s) this PR fixes:
Fixes #<105131>

```release-note
Fixed a bug that prevents PersistentVolume that has a Claim UID which doesn't exist in local cache but exists in ETCD from being updated to Released phase. 
```

### CODE REVIEW

**Comment by xiaopingrubyist** [comment]
anybody here?

File: pkg/controller/volume/persistentvolume/pv_controller.go [hunk 1]
```diff
@@ -622,9 +622,29 @@ func (ctrl *PersistentVolumeController) syncVolume(volume *v1.PersistentVolume)
 		if claim != nil && claim.UID != volume.Spec.ClaimRef.UID {
 			// The claim that the PV was pointing to was deleted, and another
 			// with the same name created.
-			klog.V(4).Infof("synchronizing PersistentVolume[%s]: claim %s has different UID, the old one must have been deleted", volume.Name, claimrefToClaimKey(volume.Spec.ClaimRef))
-			// Treat the volume as bound to a missing claim.
-			claim = nil
+			// in some cases, the cached claim is not the newest, and the volume.Spec.ClaimRef.UID is newer than cached.
+			// in this case ,we should double check by calling apiserver  to get the newest claim.
+			klog.V(3).Infof("Maybe cached claim: %s is not the newest one, we should fetch it from apiserver", claimrefToClaimKey(volume.Spec.ClaimRef))
```
Comment (msau42) [inline pkg/controller/volume/persistentvolume/pv_controller.go:627]: I would probably keep this level at 4 with the other messages here.

File: pkg/controller/volume/persistentvolume/pv_controller.go [hunk 2]
```diff
@@ -622,9 +622,29 @@ func (ctrl *PersistentVolumeController) syncVolume(volume *v1.PersistentVolume)
 		if claim != nil && claim.UID != volume.Spec.ClaimRef.UID {
 			// The claim that the PV was pointing to was deleted, and another
 			// with the same name created.
-			klog.V(4).Infof("synchronizing PersistentVolume[%s]: claim %s has different UID, the old one must have been deleted", volume.Name, claimrefToClaimKey(volume.Spec.ClaimRef))
-			// Treat the volume as bound to a missing claim.
-			claim = nil
+			// in some cases, the cached claim is not the newest, and the volume.Spec.ClaimRef.UID is newer than cached.
```
Comment (msau42) [inline pkg/controller/volume/persistentvolume/pv_controller.go:625]: I think this is basically an extension of the scenario handled at https://github.com/kubernetes/kubernetes/pull/105211/files#diff-9a553e2f939f7ad1470cfe1cf8a48b56b39c6b34fb7ad820c79f2f43602499bdR587

Is it possible to add a unit test for this scenario?

**Review by msau42 — COMMENTED** [review]
/ok-to-test

/assign @jsafrane 
for more eyes

File: pkg/controller/volume/persistentvolume/pv_controller.go (hunk 1 above)
Comment (xiaopingrubyist) [reply pkg/controller/volume/persistentvolume/pv_controller.go:627]: i will update it and back to level 4.

**Comment by brianpursley** [comment]
@xiaopingrubyist your go fmt is failing.

You can run `hack/update-gofmt.sh` to fix the issues and then it should pass.

**Comment by brianpursley** [comment]
@xiaopingrubyist also be sure to follow the PR template.

I'm just trying to help you get this PR reviewed, because it seems you have put some effort into tracking down the cause of this problem.

**Comment by xiaopingrubyist** [comment]
> @xiaopingrubyist also be sure to follow the PR template.
> 
> I'm just trying to help you get this PR reviewed, because it seems you have put some effort into tracking down the cause of this problem.

thank you so much , i will update it.

File: pkg/controller/volume/persistentvolume/pv_controller.go (hunk 2 above)
Comment (xiaopingrubyist) [reply pkg/controller/volume/persistentvolume/pv_controller.go:625]: unit test added

**Comment by jsafrane** [comment]
The code looks good now. Can you please squash the commits and provide a release note in the initial comment of this PR? This is a bug that might be interesting to some users and they should find it's fixed in the release notes.

Follow https://www.kubernetes.dev/docs/guide/release-notes/#applying-a-release-note

/triage accepted
/priority important-soon

**Comment by xiaopingrubyist** [comment]
released notes added and commits squashed. @jsafrane

**Pull request merged**